core/primitives/src/types.rs

```diff
@@ -173,6 +173,28 @@ pub struct RawStateChangesWithTrieKey {
     pub changes: Vec<RawStateChange>,
 }
 
+/// Consolidate state change of trie_key and the final value the trie key will be changed to
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize)]
+pub struct ConsolidatedStateChange {
+    pub trie_key: TrieKey,
+    pub value: Option<Vec<u8>>,
+}
+
+pub type ConsolidatedStateChanges = Vec<ConsolidatedStateChange>;
```

Comment:
nitpick: `ConsolidatedStateChange` and `ConsolidatedStateChanges` are indistinguishable at a glance. I'd just use the raw vec type.